### PR TITLE
[Calyx] Add calyx.undef

### LINKED ISSUE #5963: [Calyx] No representation for the `undef` primitive in Calyx native
The Calyx native compiler's stdlib has an `undef` value which is lowered to `'x` values. However, the semantics of `undef` are likely to be more precise than `'x` and enable optimizations that `'x` might not. We should add a new `UndefOp` to the Calyx dialect to represent this primitive.

### PR DESCRIPTION
**Pull request by rachitnigam**
Fixes #5963.

### CODE REVIEW

File: lib/Dialect/Calyx/CalyxOps.cpp [hunk 1]
```diff
@@ -2850,6 +2850,18 @@ LogicalResult SliceLibOp::verify() {
   return success();
 }
 
+// Undef op
+SmallVector<StringRef> UndefLibOp::portNames() { return {"out"}; }
+SmallVector<Direction> UndefLibOp::portDirections() { return {Output}; }
```
Comment (mortbopet) [inline lib/Dialect/Calyx/CalyxOps.cpp:2855]: ubernit: i know this is not the style used in the rest of the Calyx ops, but i would normally place such simple functions in the ODS `extraClassDeclarations` segment.

Comment (rachitnigam) [reply lib/Dialect/Calyx/CalyxOps.cpp:2855]: I noticed the lack of consistency on this too! Seems like a good idea to keep things in the `td` file as much as possible.

**Comment by rachitnigam** [comment]
@mortbopet @cgyurgyik I noticed that if I name this op `undef`, the dialect registration fails but `undefined` (or `foobar`) makes it work. I can't find anything that says that `undef` is a special identifier in MLIR dialect def so curious if y'all might know the reason for this.

**Comment by rachitnigam** [comment]
This PR ended up fixing two issues because we couldn't test the behavior for `undef` otherwise.

**Comment by rachitnigam** [comment]
I would like to merge this soon to line up with a new release of the `calyx` native compiler. I'll leave this open for another day and merge after that if folks are too busy to review

File: include/circt/Dialect/Calyx/CalyxPrimitives.td [hunk 2]
```diff
@@ -18,6 +18,38 @@ class CalyxPrimitive<string mnemonic, list<Trait> traits = []> :
     let skipDefaultBuilders = 1;
 }
 
+/// The n-bit, undef op which only provides the out signal
```
Comment (cgyurgyik) [inline include/circt/Dialect/Calyx/CalyxPrimitives.td:21]: Is this necessary?

File: include/circt/Dialect/Calyx/CalyxPrimitives.td [hunk 3]
```diff
@@ -18,6 +18,38 @@ class CalyxPrimitive<string mnemonic, list<Trait> traits = []> :
...
+def UndefLibOp: CalyxPrimitive<"undefined", []> {
+  let summary = "An undefined signal";
+  let description = [{
+    The "calyx.undef" op defines an undefined signal. An undefined signal can be
+    replaced with any value representable in n-bits.
```
Comment (cgyurgyik) [inline include/circt/Dialect/Calyx/CalyxPrimitives.td:26]: "With any n-bit value"

File: include/circt/Dialect/Calyx/CalyxPrimitives.td [hunk 4]
```diff
@@ -18,6 +18,38 @@ class CalyxPrimitive<string mnemonic, list<Trait> traits = []> :
...
+  let arguments = (ins SymbolNameAttr:$sym_name);
+  let builders = [
+    OpBuilder<(ins "StringRef":$sym_name, "::mlir::TypeRange":$resultTypes), [{
+      $_state.addAttribute(mlir::SymbolTable::getSymbolAttrName(), $_builder.getStringAttr(sym_name));
+      $_state.addTypes(resultTypes);
+    }]>
+  ];
+  let results = (outs AnyType:$out);
+  let extraClassDefinition = [{
+    bool $cppClass::isCombinational() { return false; }
```
Comment (cgyurgyik) [inline include/circt/Dialect/Calyx/CalyxPrimitives.td:41]: Why not?

File: lib/Dialect/Calyx/Export/CalyxEmitter.cpp [hunk 5]
```diff
@@ -44,7 +44,7 @@ static constexpr std::string_view exclamationMark() { return "!"; }
 static constexpr std::string_view equals() { return "="; }
 static constexpr std::string_view comma() { return ", "; }
 static constexpr std::string_view arrow() { return " -> "; }
-static constexpr std::string_view delimiter() { return "\""; }
+static constexpr std::string_view quote() { return "\""; }
```
Comment (cgyurgyik) [inline lib/Dialect/Calyx/Export/CalyxEmitter.cpp:47]: Can this be done in a follow up CL?

File: lib/Dialect/Calyx/Export/CalyxEmitter.cpp [hunk 6]
```diff
@@ -661,7 +690,7 @@ void Emitter::emitComponentPorts(ComponentInterface op) {
 
       // We only care about the bit width in the emitted .futil file.
       unsigned int bitWidth = port.type.getIntOrFloatBitWidth();
-      os << getAttributes(op, port.attributes) << port.name.getValue()
+      os << getAttributes(op, true, port.attributes) << port.name.getValue()
```
Comment (cgyurgyik) [inline lib/Dialect/Calyx/Export/CalyxEmitter.cpp:693]: /*field_name=*/true here and elsewhere

File: test/Dialect/Calyx/emit.mlir [hunk 7]
```diff
@@ -3,7 +3,7 @@
 module attributes {calyx.metadata = ["location1", "location2"], calyx.entrypoint = "main"} {
 
   // CHECK: import "primitives/core.futil";
-  // CHECK-LABEL: component A<"static"=1>(in: 32, @go go: 1, @clk clk: 1, @reset reset: 1) -> (out: 32, @done done: 1) {
+  // CHECK-LABEL: component A<"static"=1,>(in: 32, @go go: 1, @clk clk: 1, @reset reset: 1) -> (out: 32, @done done: 1) {
```
Comment (andrewb1999) [inline test/Dialect/Calyx/emit.mlir:6]: Are these commas after attributes intentional?

**Review by andrewb1999 — APPROVED** [review]
One small question, otherwise LGTM

File: test/Dialect/Calyx/emit.mlir [hunk 8]
```diff
@@ -14,7 +14,7 @@ module attributes {calyx.metadata = ["location1", "location2"], calyx.entrypoint
     calyx.control {}
   } {static = 1}
 
-  // CHECK-LABEL: component B<"toplevel"=1>(in: 1, @go go: 1, @clk clk: 1, @reset reset: 1) -> (out: 1, @done done: 1) {
+  // CHECK-LABEL: component B<"precious"=1, "toplevel"=1,>(in: 1, @go go: 1, @clk clk: 1, @reset reset: 1) -> (out: 1, @done done: 1) {
```
Comment (cgyurgyik) [inline test/Dialect/Calyx/emit.mlir:17]: We don't want a comma at the end here.

**Review by cgyurgyik — COMMENTED** [review]
A few small comments

File: test/Dialect/Calyx/emit.mlir (hunk 7 above)
Comment (rachitnigam) [reply test/Dialect/Calyx/emit.mlir:6]: Yup, unfortunately couldn't figure out a nice way to remove these. There is a `pop_back` call but I think its getting iffy to do this printing. Ideally we'd want to use `llvm::interleave` but we're also doing filtering while printing and I couldn't figure out a nice way to write that code here.

File: include/circt/Dialect/Calyx/CalyxPrimitives.td (hunk 2 above)
Comment (rachitnigam) [reply include/circt/Dialect/Calyx/CalyxPrimitives.td:21]: The doc comment? I can remove it

File: test/Dialect/Calyx/emit.mlir (hunk 8 above)
Comment (cgyurgyik) [reply test/Dialect/Calyx/emit.mlir:17]: Can't you just filter before, e.g., std::copy_if?

Comment (cgyurgyik) [reply test/Dialect/Calyx/emit.mlir:17]: This would be an ideal case for C++20 ranges but I think MLIR is still in 17 mode

Comment (cgyurgyik) [reply test/Dialect/Calyx/emit.mlir:17]: Or the llvm way is llvm::erase_if

**Pull request merged**